README.md

```diff
@@ -6,6 +6,15 @@ Express.JS frameworks.
 
 ## Features
 
+The features below are there for two basic reasons. First reason is to showcase
+some of the features of the Knative (like scaling, traffic shaping, or event
```

Comment:
Where is "scaling" or "traffic shaping" demonstrated in the showcase application, please?
If I understand it correctly this application is about sending, receiving and presenting CloudEvents, either using Node.js or Quarkus.

Reply:
There are a couple of things that help showcase Serving:

- fast app boot (requirement for Serving scale from zero),
- support for `$PORT` (automatically set by Serving),
- support for `$DELAY` (to help showcase scaling),
- support for `$GREET` (to help showcase traffic distribution, as different revisions will have different output)

> If I understand it correctly this application is about sending, receiving and presenting CloudEvents, either using Node.js or Quarkus.

No. Sending, receiving and streaming the events is just the part of the functionality.

Reply:
>     fast app boot (requirement for Serving scale from zero),
>     support for $PORT (automatically set by Serving),
>     support for $DELAY (to help showcase scaling),
>     support for $GREET (to help showcase traffic distribution, as different revisions will have different output)

OK. I get it now. Thanks! But it was not clear to me after reading the readme, just because the relation/connection between the features mentioned in readme and Knative/Serverless features is very low. The readme doesn't mention any of `$DELAY`, `$GREET` and how it's related to scaling or traffic distribution. Neither it's clear that this application will be good for demonstrating scale-from-zero because it's quarkus application and so it starts quickly.

If I'm seeing this application and the source code for the first time, I'd expect something like this in the readme:
* The application is feasible for scaling-from-zero because it's a quarkus/expressjs app with a quick startup 
* The application can be configured to set delay on requests so that parallel requests to the application will cause it to scale up (plus example values/steps):
   `kn service create showcase --image quay.io/openshift-knative/showcase --env DELAY=5000`
* The application can be configured to provide different responses for different versions of application which can demonstrate creating new revisions:
   `kn service create showcase --image quay.io/openshift-knative/showcase --env GREET=Hello` followed by `kn service update showcase --env GREET=Hi`
(traffic splitting requires additional configuration)

I'm still missing how PORT is related to "Knative features" but it could probably be described in the readme as well.

Overall, somebody who knows the purpose of the application from the beginning will know how to use those features in connection with Knative. I just want to make it clear for new people seeing this app. After all, it's an app for demonstration purposes.

Reply:
Since this application is useful for presenters (as you mentioned in [here](https://github.com/openshift-knative/showcase/pull/11#discussion_r1164634346)). Let's make it clear what this app offers in the readme (let's list those features). I.e. be more specific in how these features are related to Knative and how they can be used to demonstrate Knative features. It shouldn't be required to read every line of the code to understand that.

Reply:
> The readme doesn't mention any of `$DELAY`, `$GREET`

It mentions the `DELAY`. `GREET` isn't mentioned, however :crying_cat_face: 

> I'm still missing how `PORT` is related to "Knative features"

Regular K8s doesn't provide the PORT env variable to the apps. People need to specify the `containerPort` to match their app. With Knative this is automatic, as long as app respects the `PORT` given by Knative. Here's the support in Knative AFAIK: https://github.com/knative/serving/blob/018c7f81957bbe5d152d6a0b5569c12e2baab0bd/pkg/reconciler/revision/resources/deploy.go#L328-L333

That's why this is strictly Knative related feature.

---

All in all, I like your explanations very much. They are superior. We definitely should improve the README to make it clear.

However, I would like to do it in separate follow-up PR. There are a couple of things we can improve in follow-up PRs, in a parallel way: build improvements, README improvements, Quarkus streaming threads fixes, event logging style.

Merging this PR will unblock the docs team, so we can start incorporating the showcase image in the OCP docs, and we can improve the app in smaller focused PRs.

WDYT?

Reply:
> Regular K8s doesn't provide the PORT env variable to the apps. People need to specify the containerPort to match their app. With Knative this is automatic, as long as app respects the PORT given by Knative. Here's the support in Knative AFAIK: https://github.com/knative/serving/blob/018c7f81957bbe5d152d6a0b5569c12e2baab0bd/pkg/reconciler/revision/resources/deploy.go#L328-L333

Yeah. OK. This application in the current shape can run on a specific port according to the PORT variable set on the Pod. For KSVC it's enough to set 
```
  spec:
    template:
      spec:
        containers:
          ports:
          - containerPort: XYZ
            protocol: TCP
``` 
and the PORT env variable will be added automatically to the Pod.
For regular deployment it's required to add the same and also add:
```
  - env:
    - name: PORT
      value: XYZ
```
Got it. This is just for completeness. Thanks

Reply:
> However, I would like to do it in separate follow-up PR. There are a couple of things we can improve in follow-up PRs, in a parallel way: build improvements, README improvements, Quarkus streaming threads fixes, event logging style.

OK. It's alright to do this in follow up PRs.

Reply:
Hmm, I doubt so. How would Knative know what port your custom app image runs on? Knative adds the PORT env variable based on "spec.container.port" of the given revision. Someone has to specify the different port, like in the case below:
```
❯ kn service create hello2 --image gcr.io/knative-samples/helloworld-go --port 8083 --env TARGET=World
Creating service 'hello2' in namespace 'default':

  0.118s Configuration "hello2" is waiting for a Revision to become ready.
  4.226s ...
  4.251s Ingress has not yet been reconciled.
  4.327s Waiting for load balancer to be ready
  4.501s Ready to serve.

Service 'hello2' created to latest revision 'hello2-00001' is available at URL:
https://hello2-default.apps.mgencur-381.ci-psi.rhocf-dev.net



❯ oc get ksvc hello2 -oyaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  annotations:
    serving.knative.dev/creator: kube:admin
    serving.knative.dev/lastModifier: kube:admin
  creationTimestamp: "2023-04-13T10:35:20Z"
  generation: 1
  name: hello2
  namespace: default
  resourceVersion: "669207"
  uid: ba28cb0d-e629-404c-b5e5-ddd255c99a33
spec:
  template:
    metadata:
      annotations:
        client.knative.dev/updateTimestamp: "2023-04-13T10:35:19Z"
        client.knative.dev/user-image: gcr.io/knative-samples/helloworld-go
      creationTimestamp: null
    spec:
      containerConcurrency: 0
      containers:
      - env:
        - name: TARGET
          value: World
        image: gcr.io/knative-samples/helloworld-go
        name: user-container
        ports:
        - containerPort: 8083
          protocol: TCP
        readinessProbe:
          successThreshold: 1
          tcpSocket:
            port: 0
        resources: {}
      enableServiceLinks: false
      timeoutSeconds: 300
  traffic:
  - latestRevision: true
    percent: 100
status:
  address:
    url: http://hello2.default.svc.cluster.local
  conditions:
  - lastTransitionTime: "2023-04-13T10:35:24Z"
    status: "True"
    type: ConfigurationsReady
  - lastTransitionTime: "2023-04-13T10:35:25Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2023-04-13T10:35:25Z"
    status: "True"
    type: RoutesReady
  latestCreatedRevisionName: hello2-00001
  latestReadyRevisionName: hello2-00001
  observedGeneration: 1
  traffic:
  - latestRevision: true
    percent: 100
    revisionName: hello2-00001
  url: https://hello2-default.apps.mgencur-381.ci-psi.rhocf-dev.net
```

Reply:
Here you have an old version of this app, which runs by default on port `21111`: https://github.com/cardil/knative-serving-showcase/blob/1207fc00a96b3ba058ab2fc47d07a52ae22f4a55/src/main/templates/application.properties#L2

You can deploy that app with just:

```shell
$ kn service create showcase --image quay.io/cardil/knative-serving-showcase
```

It will run on Knative, as Knative will set the `PORT==8080` by default. Because this app respects the `PORT` it will start on that port. So, you can just run the image, without any port parameters explicitly set.

Reply:
BTW2. The image you gave as an example, also supports the `PORT`: https://github.com/knative-sample/helloworld-go/blob/73fbdd56107f593bd6cf8852a5649276540f4340/helloworld.go#L41

Reply:
BTW3. In your example, see the env variables that Knative set on the app container. They contain the PORT, despite you have given such env var.